packages/assets-controllers/src/TokenDetectionController.ts

```diff
@@ -1,242 +1,274 @@
+  readonly #getTokensState: () => TokensState;
 
   /**
    * Creates a TokenDetectionController instance.
    *
    * @param options - The controller options.
+   * @param options.messenger - The controller messaging system.
+   * @param options.interval - Polling interval used to fetch new token rates
+   * @param options.chainId - The chain ID of the current network
+   * @param options.selectedAddress - Vault selected address
    * @param options.onPreferencesStateChange - Allows subscribing to preferences controller state changes.
-   * @param options.onNetworkStateChange - Allows subscribing to network controller state changes.
-   * @param options.onTokenListStateChange - Allows subscribing to token list controller state changes.
    * @param options.getBalancesInSingleCall - Gets the balances of a list of tokens for the given address.
    * @param options.addDetectedTokens - Add a list of detected tokens.
-   * @param options.getTokenListState - Gets the current state of the TokenList controller.
    * @param options.getTokensState - Gets the current state of the Tokens controller.
-   * @param options.getNetworkState - Gets the state of the network controller.
    * @param options.getPreferencesState - Gets the state of the preferences controller.
-   * @param options.getNetworkClientById - Gets the network client by ID.
-   * @param config - Initial options used to configure this controller.
-   * @param state - Initial state to set on this controller.
    */
-  constructor(
-    {
-      onPreferencesStateChange,
-      onNetworkStateChange,
-      onTokenListStateChange,
-      getBalancesInSingleCall,
-      addDetectedTokens,
-      getTokenListState,
-      getTokensState,
-      getNetworkState,
-      getPreferencesState,
-      getNetworkClientById,
-    }: {
-      onPreferencesStateChange: (
-        listener: (preferencesState: PreferencesState) => void,
-      ) => void;
-      onNetworkStateChange: (
-        listener: (networkState: NetworkState) => void,
-      ) => void;
-      onTokenListStateChange: (
-        listener: (tokenListState: TokenListState) => void,
-      ) => void;
-      getBalancesInSingleCall: AssetsContractController['getBalancesInSingleCall'];
-      addDetectedTokens: TokensController['addDetectedTokens'];
-      getTokenListState: () => TokenListState;
-      getTokensState: () => TokensState;
-      getNetworkState: () => NetworkState;
-      getPreferencesState: () => PreferencesState;
-      getNetworkClientById: NetworkController['getNetworkClientById'];
-    },
-    config?: Partial<TokenDetectionConfig>,
-    state?: Partial<BaseState>,
-  ) {
-    const {
-      providerConfig: { chainId: defaultChainId },
-    } = getNetworkState();
+  constructor({
+    chainId,
+    selectedAddress = '',
+    interval = DEFAULT_INTERVAL,
+    onPreferencesStateChange,
+    getBalancesInSingleCall,
+    addDetectedTokens,
+    getTokensState,
+    getPreferencesState,
+    messenger,
```

Comment:
@MajorLift I think this is the only comment still unresolved. Just highlighting to make sure you saw it.

Reply:
Applied here: 5fe76743e6bac267557778baaede62ce43c2fccd

I'm a little wary of providing this option. It might add unnecessary complexity by allowing multiple ways of arriving at some states. I feel like minimizing the surface area of the API by only exposing `{start,stop}{,PollingByNetworkClientId}` might be less implicit and make for more readable code. If I'm understanding this correctly, the method-state mappings we have now are pretty tricky:

- injected network service
`enableNetwork`: `(polling) => { networkDisabled: false, polling }`
`disableNetwork` : `() => { networkDisabled: true, polling: false }`
- polling controller
`init`, `stopPolling`: `(networkDisabled) => { networkDisabled, polling: false }`
`startPolling`: `(networkDisabled) => { networkDisabled, polling: !networkDisabled }` (fails if networkDisabled is true)
`start`: `() => { networkDisabled: false, polling: true }`
`stop`: `() => { networkDisabled: true, polling: false }`

But since it's unclear how interactions with the service object will be implemented, I'm good with leaving this option as it is now and refactoring down the road.